### ISSUE TITLE
Using start TransferService is Application Class Also for a background event crashes in Android O & above

### ISSUE THREAD

**Issue by deepakrokz** [aws-amplify/aws-sdk-android#1110]
**Describe the bug**
Here is in my application, I am trying to start TransferService in Application class but it's being crashed all the time.

**To Reproduce**
A code sample or steps:
```
 if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                try {
                    mInstance.getApplicationContext().startForegroundService(new Intent(mInstance.getApplicationContext(), TransferService.class));
                } catch (Exception e) {
                    e.printStackTrace();
                }
}
```

**Which AWS service(s) are affected?**

2.14.1 and older

**Expected behavior**

**Screenshots**

**Environment Information (please complete the following information):**
 - 2.14.1
 - Device: Vivo and others running on/above oreo
 - Android Version: oreo & above
 - Specific to simulators: Not sure

**Additional context**
I was earlier starting this service only with startService but started getting crashes in several devices that service can't be started in the background. Please help me out also with earlier startService, the file wasn't being uploaded on S3 while the app is in background.

Thanks in advance.

**Comment by kvasukib** [comment]
Calling `startService` when the app is in the background will crash your application on devices running Oreo and above. This is due to the limitations on background processing that Android has. 

The solution for devices running Oreo and above would be to invoke `startForegroundService` with a valid notification object passed through the intent to `startForegroundService`.

Please refer to [Supporting TransferService on Oreo and above](https://aws-amplify.github.io/docs/android/storage#supporting-transferservice-on-oreo-and-above) for additional details.

**Comment by deepakrokz** [comment]
Hi @kvasukib,

I have given my sample code of starting service that's what I am using and I.e. startForegroundService not startService. Please suggest

**Comment by deepakrokz** [comment]
I know calling startService in background will crash the app for running os oreo and above I have already mentioned

**Comment by kvasukib** [comment]
@deepakrokz Thank you for the information, You need to pass in a valid notification object through the intent to `startForegroundService`. Are you passing one? I couldn't find that in your sample code.

The [documentation for TransferService](https://aws-amplify.github.io/docs/android/storage#supporting-transferservice-on-oreo-and-above) talks about it.

What is the crash you are experiencing? Can you post the stacktrace?

**Comment by deepakrokz** [comment]
Hi, I have simply started the foregroundservice without any notification id. I will share crash report in sometime

**Comment by deepakrokz** [comment]
This is the crash log 
 android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{45f5c55 u0 com.canlah/com.amazonaws.mobileconnectors.s3.transferutility.TransferService}
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1760)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:201)
        at android.app.ActivityThread.main(ActivityThread.java:6810)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:547)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:873)
 E/UncaughtException: android.app.RemoteServiceException: Context.startForegroundService() did not then call Service.startForeground(): ServiceRecord{45f5c55 u0 com.canlah/com.amazonaws.mobileconnectors.s3.transferutility.TransferService}
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1760)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:201)
        at android.app.ActivityThread.main(ActivityThread.java:6810)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:547)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:873)

**Comment by kvasukib** [comment]
@deepakrokz This is due to the fact that you did not pass a valid notification object through the intent. See [my above comment ](https://github.com/aws-amplify/aws-sdk-android/issues/1110#issuecomment-518096863)about how to pass in a valid notification through the intent. 

On devices running Oreo and above, a notification needs to be displayed in the system tray to run the service in foreground. The SDK expects the app to pass in a notification through the intent.

**Comment by deepakrokz** [comment]
@kvasukib thanks for your instant support. I tried your notification related intents but not exactly able to configure intent fields values. Can you please help me with that so that I can test and finish this. There is no such example available out there in my search

**Comment by kvasukib** [comment]
Th [documentation](https://aws-amplify.github.io/docs/android/storage#supporting-transferservice-on-oreo-and-above) has the example of how to pass a notification through the intent.

I am pasting the code snippet again here for reference.

```
Intent tsIntent = new Intent(getApplicationContext(), TransferService.class);
tsIntent.putExtra(TransferService.INTENT_KEY_NOTIFICATION, <notification-object>);
tsIntent.putExtra(TransferService.INTENT_KEY_NOTIFICATION_ID, <notification-id>);
tsIntent.putExtra(TransferService.INTENT_KEY_REMOVE_NOTIFICATION, <remove-notification-when-service-stops-foreground>);
getApplicationContext().startForegroundService(tsIntent);
```

**Comment by deepakrokz** [comment]
Hi @kvasukib,
I have seen this and tried to pass different string values for each putExtra keys, but it's not satisfying. That's why I asked for any configured example of this with some real data. I will modify that according to my need. Waiting for your quick response on the same.
Thanks

**Comment by deepakrokz** [comment]
`Intent tsIntent = new Intent(mInstance.getApplicationContext(), TransferService.class);
        NotificationChannel channel = new NotificationChannel(CHANNEL_MESSAGE,
                "Channel human readable title",
                NotificationManager.IMPORTANCE_NONE);
        ((NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE)).createNotificationChannel(channel);

        Notification notification = new NotificationCompat.Builder(this, CHANNEL_MESSAGE)
                .setContentTitle("")
                .setContentText("").build();

        tsIntent.putExtra(TransferService.INTENT_KEY_NOTIFICATION, notification);
        tsIntent.putExtra(TransferService.INTENT_KEY_NOTIFICATION_ID, 113);
        tsIntent.putExtra(TransferService.INTENT_KEY_REMOVE_NOTIFICATION, true);
        mInstance.getApplicationContext().startForegroundService(tsIntent);
`
I have tried this way as suggested but it's creating a notification in Notification Panel, I don't want something like also when removing the app from the recent app list. The service is being closed. How can I keep this running even if the app is closed?

Please help me

**Comment by deepakrokz** [comment]
Hi @kvasukib,

Please respond, it's really urgent. Also one request why don't you ask your team to update sample related to S3 on github. It will really helpful for everyone.

Waiting for your reply


Thanks

**Comment by kvasukib** [comment]
@deepakrokz It is expected for the notification to show up in the system tray when the service is running. This is the expected behavior of Android Foreground services. See https://developer.android.com/guide/components/services#Foreground for more information.

Our SDK (TransferUtility) currently does not support running the transfers when the app is closed. We will evaluate if we want to support this feature.

For the sample app update, we will work on updating our sample app with an example of notification and starting foreground service. Thank you for pointing it out.

**Comment by deepakrokz** [comment]
Hi @kvasukib,

Thanks for getting back, actually we don't want to continuously display a notification to users as it will be more irritating for the user. Looking for background stuff in case of app killed 

Thanks for your support.

**Comment by sictiru** [comment]
+1 on updating the sample app. Just started on changing the S3 SDK from an older version but found the docs I need here.

**Comment by palpatim** [comment]
We will take on updating the sample app as a feature request and update here when we have more information.

**Comment by deepakrokz** [comment]
Hi @kvasukib, 

Please update eagerly waiting for your update..

**Comment by raphkim** [comment]
Duplicate of closed issue #1173 by the same requester
